File: ObjC/CGPathUnionObjC/CGPathUnionObjC/PathGen.m
//
//  PathGen.m
//  CGPathUnionObjC
//
//  Created by Don Mag on 6/30/23.
//

#import "PathGen.h"

@implementation PathGen

+ (CGMutablePathRef)buildShape:(ShapeType)s rect:(CGRect)rect {
	return [self buildShape:s rect:rect numSides:0];
}
+ (CGMutablePathRef)buildShape:(ShapeType)s rect:(CGRect)rect numSides:(NSInteger)numSides {
	CGMutablePathRef newPath = CGPathCreateMutable();
	
	switch (s) {
		case kSquare:
			newPath = CGPathCreateMutableCopy(CGPathCreateWithRect(rect, nil));
			break;
			
		case kCircle:
			newPath = CGPathCreateMutableCopy(CGPathCreateWithEllipseInRect(rect, nil));
			break;
			
		case kDiamond:
			CGPathMoveToPoint(newPath, nil, CGRectGetMidX(rect), CGRectGetMinY(rect));
			CGPathAddLineToPoint(newPath, nil, CGRectGetMaxX(rect), CGRectGetMidY(rect));
			CGPathAddLineToPoint(newPath, nil, CGRectGetMidX(rect), CGRectGetMaxY(rect));
			CGPathAddLineToPoint(newPath, nil, CGRectGetMinX(rect), CGRectGetMidY(rect));
			CGPathCloseSubpath(newPath);
			break;
			
		case kTriangle:
			CGPathMoveToPoint(newPath, nil, CGRectGetMinX(rect), CGRectGetMaxY(rect));
			CGPathAddLineToPoint(newPath, nil, CGRectGetMidX(rect), CGRectGetMinY(rect));
			CGPathAddLineToPoint(newPath, nil, CGRectGetMaxX(rect), CGRectGetMaxY(rect));
			CGPathCloseSubpath(newPath);
			break;
			
		case kPacMan:
		{
			CGRect r = rect;
			r.origin = CGPointZero;
			r.size.height = r.size.width;
			CGPoint cntr = CGPointMake(CGRectGetMidX(r), CGRectGetMidY(r));
			CGPathMoveToPoint(newPath, nil, cntr.x, cntr.y);
			CGPathAddArc(newPath, nil, cntr.x, cntr.y, r.size.width * 0.5, M_PI * 0.25, M_PI * 1.75, NO);
			CGPathCloseSubpath(newPath);
			CGAffineTransform t1 = CGAffineTransformScale(CGAffineTransformIdentity, 1.0, rect.size.height / rect.size.width);
			CGAffineTransform t2 = CGAffineTransformTranslate(t1, rect.origin.x, rect.origin.y);
			newPath = CGPathCreateMutableCopyByTransformingPath(newPath, &t2);
		}
			break;
			
		case kPolygon:
			if (numSides < 1) break;
		{
			// we'll base the polygon on a 1:1 circle
			CGRect r = rect;
			r.origin = CGPointZero;
			r.size.height = r.size.width;
			CGFloat radius = r.size.width * 0.5;
			CGPoint cntr = CGPointMake(CGRectGetMidX(r), CGRectGetMidY(r));
			// we *could* use maths to calculate the polygon points, but instead
			//	let's "cheat" by adding arcs, and get the points from each arc endPoint
			CGMutablePathRef arcPath = CGPathCreateMutable();
			CGMutablePathRef ptPath = CGPathCreateMutable();
			// start at "12 o'clock"
			double startA = M_PI * 1.5;
			double radianInc = (M_PI * 2.0) / (double)numSides;
			CGPoint pt = CGPointMake(CGRectGetMidX(r), CGRectGetMinY(r));
			CGPathMoveToPoint(arcPath, nil, pt.x, pt.y);
			CGPathMoveToPoint(ptPath, nil, pt.x, pt.y);
			for (int i = 0; i < (numSides - 1); i++) {
				CGPathAddArc(arcPath, nil, cntr.x, cntr.y, radius, startA, startA + radianInc, YES);
				CGPoint curPT = CGPathGetCurrentPoint(arcPath);
				CGPathAddLineToPoint(ptPath, nil, curPT.x, curPT.y);
				startA += radianInc;
			}
			CGPathCloseSubpath(ptPath);
			// now we need to scale the polygon to the desired rect
			CGAffineTransform t1 = CGAffineTransformScale(CGAffineTransformIdentity, 1.0, rect.size.height / rect.size.width);
			CGAffineTransform t2 = CGAffineTransformTranslate(t1, rect.origin.x, rect.origin.y);
			newPath = CGPathCreateMutableCopyByTransformingPath(ptPath, &t2);
		}
			break;
			
		default:
			break;
	}
	
	return newPath;
}

@end

File: Swift/CGPathUnionSwift/CGPathUnionSwift/PathGen.swift
//
//  PathGen.swift
//  CGPathUnionSwift
//
//  Created by Don Mag on 6/30/23.
//

import UIKit

enum ShapeType : Int {
	case kSquare, kDiamond, kTriangle, kCircle, kPacMan, kPolygon
}

class PathGen: NSObject {

	func buildShape(_ s: ShapeType, in rect: CGRect) -> CGMutablePath {
		buildShape(s, in: rect, numSides: 0)
	}
	func buildShape(_ s: ShapeType, in rect: CGRect, numSides: Int) -> CGMutablePath {
		var newPath = CGMutablePath()
		switch s {
		case .kSquare:
			newPath = CGMutablePath(rect: rect, transform: nil)
			
		case .kCircle:
			newPath = CGMutablePath(ellipseIn: rect, transform: nil)
			
		case .kDiamond:
			newPath.move(to: .init(x: rect.midX, y: rect.minY))
			newPath.addLine(to: .init(x: rect.maxX, y: rect.midY))
			newPath.addLine(to: .init(x: rect.midX, y: rect.maxY))
			newPath.addLine(to: .init(x: rect.minX, y: rect.midY))
			newPath.closeSubpath()
			
		case .kTriangle:
			newPath.move(to: .init(x: rect.minX, y: rect.maxY))
			newPath.addLine(to: .init(x: rect.midX, y: rect.minY))
			newPath.addLine(to: .init(x: rect.maxX, y: rect.maxY))
			newPath.closeSubpath()
			
		case .kPacMan:
			var r: CGRect = rect
			r.origin = .zero
			r.size.height = r.size.width
			let cntr: CGPoint = .init(x: r.midX, y: r.midY)
			newPath.move(to: cntr)
			newPath.addArc(center: cntr, radius: r.width * 0.5, startAngle: .pi * 0.25, endAngle: .pi * 1.75, clockwise: false)
			newPath.closeSubpath()
			var tr = CGAffineTransform(scaleX: 1.0, y: rect.height / rect.width)
				.concatenating(.init(translationX: rect.origin.x, y: rect.origin.y))
			if let trPath = newPath.mutableCopy(using: &tr) {
				newPath = trPath
			}

		case .kPolygon:
			var r: CGRect = rect
			r.origin = .zero
			r.size.height = r.size.width
			let cntr: CGPoint = .init(x: r.midX, y: r.midY)
			let ptPath = CGMutablePath()
			let sPath = CGMutablePath()
			let radInc: CGFloat = (.pi * 2.0) / CGFloat(numSides)
			var startA: CGFloat = .pi * 1.5
			let pt: CGPoint = .init(x: r.midX, y: r.minY)
			ptPath.move(to: pt)
			sPath.move(to: pt)
			for _ in 0..<(numSides - 1) {
				ptPath.addArc(center: cntr, radius: r.size.width * 0.5, startAngle: startA, endAngle: startA + radInc, clockwise: true)
				sPath.addLine(to: ptPath.currentPoint)
				startA += radInc
			}
			sPath.closeSubpath()
			var tr = CGAffineTransform(scaleX: 1.0, y: rect.height / rect.width)
				.concatenating(.init(translationX: rect.origin.x, y: rect.origin.y))
			if let trPath = sPath.mutableCopy(using: &tr) {
				newPath = trPath
			}
			
		default:
			()
		}
		return newPath
	}

}
